docs/design/features/greenthreads.md

```diff
@@ -0,0 +1,233 @@
+mov     eax, 0x34000008    // Argument to _morestack function describing incoming argument
+                           // size and stack space needed for function
+call   _morestack          // Call to _morestack function to jump to new stack
+NormalProlog :             // Label indicating the start of the normal prolog of the method
+```
+
+#### P/Invoke transitions
+As normal C code cannot safely run in a green stack segment, all p/invokes must transition to a normal thread. This transition adds significant code to each p/invoke transition. For example, consider the following 2 segments of assembly. The first is without green thread transition support, and second is with.
+
+##### P/Invoke transition without green thread support
+```
+lea     rcx, [rbp - 144]                                   ; Perform GC transition to GC Preemptive
+call    qword ptr [0xba67e8]  // PINVOKE_BEGIN (HELPER) 
+mov     rax, qword ptr [0xbc6650] (INDIRECT_PINVOKE_TARGET); Load address of native function
+mov     rax, qword ptr [rax]
+mov     rdx, rsi
+xor     ecx, ecx
+mov     r9d, 2
+call    rax                                    ; Call the native function, or thunk to OS thread function
+mov     esi, eax                               ; Handle return value
+lea     rcx, [rbp - 144]                       ; Perform GC transition back to GC Cooperative
+call    qword ptr [0xba67f0]  // PINVOKE_END (HELPER)
+```
+
+##### P/Invoke transition with green thread support
+```
+lea     rcx, [rbp - 144]                                   ; Perform GC transition to GC Preemptive
+call    qword ptr [0xba67e8]  // PINVOKE_BEGIN (HELPER) 
+mov     rcx, qword ptr [0xbc6650] (INDIRECT_PINVOKE_TARGET); Load address of native function
+mov     rcx, qword ptr [rcx]
+mov     edx, 32                                       ; Size of stack arguments needed by target function
+; Call helper function which if on a green thread will
+; setup for a green thread transition, and replace
+; the function pointer with a transition thunk
+call    qword ptr [0xba68d0]  // COMPUTE_GREEN_THREAD_TRANSITION 
+mov     r8d, edi                                        ; Load register arguments for calling native function
+mov     rdx, rsi
+xor     ecx, ecx
+mov     r9d, 2
+call    rax                                     ; Call the native function, or thunk to OS thread function
+mov     esi, eax                                ; Handle return value
+lea     rcx, [rbp - 144]                        ; Perform GC transition back to GC Cooperative
+call    qword ptr [0xba67f0]                    ; PINVOKE_END (HELPER)
+```
+
+#### Stack layout diagrams
+##### Diagram of what the stack may look like when using multiple stack segments and a pinvoke.
+![ExperimentalStackLayout](greenthreads_diagrams/ExperimentalStackLayout.svg)
+ 
+
+##### Before yield/after yield
+
+![YieldOperation](greenthreads_diagrams/YieldOperation.svg)
+![YieldOperation2](greenthreads_diagrams/YieldOperation2.svg)
```

Comment:
```suggestion

Diagram of what the stack may look like when using multiple stack segments and a p/invoke:

![ExperimentalStackLayout](greenthreads_diagrams/ExperimentalStackLayout.svg)
 

Diagram of before yield:

![YieldOperation](greenthreads_diagrams/YieldOperation.svg)

Diagram of after yield:

![YieldOperation2](greenthreads_diagrams/YieldOperation2.svg)
```

Reply:
Are these images actually different? It seems like two screen caps of the same slide.

Reply:
They are different now.